### ISSUE TITLE
App crashes at start

### ISSUE THREAD

**Issue by drmiaji** [RMA-17/MyQoran-Compose#1]
Dear sir, this app is very valuable. But it crashes at start. It says, 

"java.lang.IllegalArgumentException: Expected URL scheme 'http' or 'https' but no scheme was found for null"

Could you please have a look at it? Thanks in advance.

**Comment by rmaprojects** [comment]
Of course it crashes. Because you built it from it's sourcecode. The URL scheme mentioned is hidden.
If you take a look at the [Injection Section](https://github.com/rmaprojects/MyQoran-Compose/blob/master/app/src/main/java/com/rmaproject/myqoran/di/AppModule.kt):
```kotlin
@Provides
@Singleton
fun provideApiConfig(): ApiInterface {
     //....
     return Retrofit.Builder()
         .baseUrl(BuildConfig.API_BASE_URL) // This line
         .client(client)
         .addConverterFactory(GsonConverterFactory.create())
         .build()
         .create(ApiInterface::class.java)
}
```
The BuildConfig.API_BASE_URL contains nothing. It originates from [build.gradle](https://github.com/rmaprojects/MyQoran-Compose/blob/master/app/build.gradle) section that grabbed the URL from local.properties:

```groovy
plugins {
    id 'com.android.application'
    id 'org.jetbrains.kotlin.android'
    id 'com.google.dagger.hilt.android'
    id 'com.google.devtools.ksp'
    id 'com.google.gms.google-services'
    id 'com.google.firebase.crashlytics'
    id 'com.google.firebase.firebase-perf'
}

android {
    namespace 'com.rmaproject.myqoran'
    compileSdk 34

    defaultConfig {
      //...
        
       // These lines:
        Properties properties = new Properties()
      // It grabbed local.properties. local.properties file is in my local machine which is secret
        properties.load(project.rootProject.file("local.properties").newDataInputStream())

        buildConfigField "String", "AUDIO_BASE_URL", "\"${properties.getProperty("AUDIO_BASE_URL")}\""
        buildConfigField "String", "NOTIFICATION_ICON_URL", "\"${properties.getProperty("NOTIFICATION_ICON_URL")}\""
        buildConfigField "String", "API_BASE_URL", "\"${properties.getProperty("API_BASE_URL")}\""
        buildConfigField "String", "QIBLA_FINDER_URL_ID", "\"${properties.getProperty("QIBLA_FINDER_URL_ID")}\""
        buildConfigField "String", "QIBLA_FINDER_URL_EN", "\"${properties.getProperty("QIBLA_FINDER_URL_EN")}\""
    }

   //...
}
```

If you want to try my app you can find [my app on Google Play Store](https://play.google.com/store/apps/details?id=com.rmaproject.myqoran)

**Comment by drmiaji** [comment]
Yes, I tried your app and found it very interesting. How can I reuse your
codes for my own translation of Qoran? I'm doing this fi sabilillah. My old
app has 50k downloads and needs to be updated. The features you used are
very nice. I liked :)

I thank you in advance and wish you all the best. May Allah Subhanhu TaAla
give you rewards in this life and hereafter. Ameen.

Abdul-Baten Miaji
Systems Engineer
Wipro Scandi

**Comment by rmaprojects** [comment]
Well, If you use the same Framework as Mine (Jetpack Compose) you can easily transfer/copy. my code to your own. If you are still using the old way to create Android Apps (with Views XML) you have to migrate or build it from start again using Jetpack Compose. You can always start your new Android Jetpack Compose project from Android Studio. Just update it into the latest version (iguana) and select the "Empty Activity"
<br/>
<img width="900" alt="Screenshot 2024-04-08 at 09 07 26" src="https://github.com/rmaprojects/MyQoran-Compose/assets/69703296/070004c7-5595-4ae5-bec8-8cb71dc23a9e">

**Comment by drmiaji** [comment]
Thanks again for your time and response. I'm not expert on this. You did
excellent job and migrated the whole project to the jetpack compose. It's
useful. I'm grateful for your permission to use your codes. It saves my
time.

Now I wonder how to bypass urls or how to get it to work? I'm going to
acknowledge your efforts in my project. Again thanks a lot and BarakAlkahu
fik.

Abdul-Baten Miaji
Systems Engineer
Wipro Scandi

**Comment by drmiaji** [comment]
Alhamdulillah, I've successfully run the app. How to handle the sqlite
database update function? This function is missing. Could you please update
this subclass? I'm acknowledge your hardwork in my app. What name should I
put? RMA project only or the name of developer? Many many thanks. May Allah
bless you.

Abdul-Baten Miaji
Systems Engineer
Wipro Scandi

**Comment by rmaprojects** [comment]
@DrMiaji IIRC, My app doesn't have any UPDATE function for Qur'an, how can you update the Qur'an 😅? The database functions are only for retrieving ayahs and sorahs, so no UPDATE option for Qur'an. And yes, you can mention me as **RMA Projects** as the name of the Developer. May Allah bless you for your efforts

**Comment by drmiaji** [comment]
Thank you for your response. Eid Mubarak!

Abdul-Baten Miaji
Systems Engineer
Wipro Scandi

**Comment by drmiaji** [comment]
Dear brother, after working on it for several weeks, now I have found the issue and resolved it. 

I removed the @DatabaseView("SELECT id, sora, sora_name_ar, sora_name_en, sora_name_id, COUNT(id) as ayah_total, sora_descend_place FROM quran GROUP by sora") to the QoranDao. 

The secone thing I changed is AppModule file and QoranDatabse files. 

In AppModule file I kept only following: 
@Provides
    @Singleton
    fun provideQuranDatabase(app: Application): QuranDatabase {
        return QuranDatabase.getInstance(app)
    }

In QoranDatabse file I added the following: 

@Database(
    version = 5, entities = [Quran::class],
    // views = [Surah::class, Juz::class, Page::class, Surah::class],

    exportSchema = true, 
)
abstract class QuranDatabase : RoomDatabase() {
    abstract fun quranDao(): QuranDao

    companion object {
        @Volatile
        private var INSTANCE: QuranDatabase? = null
        fun getInstance(context: Context): QuranDatabase {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: Room.databaseBuilder(
                    context,
                    QuranDatabase::class.java,
                    "qoran.db"
                ).createFromInputStream {
                    context.resources.openRawResource(R.raw.qoran)
                }
                    .fallbackToDestructiveMigration()                 
                    .build()
            }
        }
    }
}

**Comment by drmiaji** [comment]
.fallbackToDestructiveMigration() is now working fine to delete the old database and add the new one. I am still working on it to make it work with Search functions so that both surah and ayah search are in the same location. If you have any update please let me know.

**Comment by drmiaji** [comment]
Dear brother, assalamu alaikum warahmatullah! Hope you are doing well by the grace of Almighty Allah. Alhamdulillah, I am fine. 

On the day of Arafah, near the Arafah mountain, I prayed especially for you as I borrowed your idea and developed my Quran app. May Allah accept my payer and dua. If you have any update, I would be very grateful if you please share the update with me. BarakAllahu fikum.
